Mathlib/Data/Fin/VecNotation.lean

```diff
@@ -0,0 +1,590 @@
+
+-- Porting note: the next two decls are commented out. TODO(eric-wieser)
+
+-- /- ./././Mathport/Syntax/Translate/Tactic/Builtin.lean:76:14:
+--  unsupported tactic `reflect_name #[] -/
+-- /- ./././Mathport/Syntax/Translate/Tactic/Builtin.lean:76:14:
+--  unsupported tactic `reflect_name #[] -/
+-- unsafe instance _root_.pi_fin.reflect [reflected_univ.{u}] [reflected _ α] [has_reflect α] :
+--     ∀ {n}, has_reflect (Fin n → α)
+--   | 0, v =>
+--     (Subsingleton.elim vecEmpty v).rec
+--       ((by
+--             trace
+--               "./././Mathport/Syntax/Translate/Tactic/Builtin.lean:76:14:
+--                unsupported tactic `reflect_name #[]" :
+--             reflected _ @vecEmpty.{u}).subst
+--         q(α))
+--   | n + 1, v =>
+--     (cons_head_tail v).rec <|
+--       (by
+--             trace
+--               "./././Mathport/Syntax/Translate/Tactic/Builtin.lean:76:14:
+--                unsupported tactic `reflect_name #[]" :
+--             reflected _ @vecCons.{u}).subst₄
+--         q(α) q(n) q(_) (_root_.pi_fin.reflect _)
+-- #align pi_fin.reflect pi_fin.reflect
+
+-- /-- Convert a vector of pexprs to the pexpr constructing that vector.-/
+-- unsafe def _root_.pi_fin.to_pexpr : ∀ {n}, (Fin n → pexpr) → pexpr
+--   | 0, v => ``(![])
+--   | n + 1, v => ``(vecCons $(v 0) $(_root_.pi_fin.to_pexpr <| vecTail v))
+-- #align pi_fin.to_pexpr pi_fin.to_pexpr
+
+/-! ### Numeral (`bit0` and `bit1`) indices
+The following definitions and `simp` lemmas are to allow any
+numeral-indexed element of a vector given with matrix notation to
+be extracted by `simp` (even when the numeral is larger than the
+number of elements in the vector, which is taken modulo that number
+of elements by virtue of the semantics of `bit0` and `bit1` and of
+addition on `fin n`).
+-/
+
+
+/-- `vecAppend ho u v` appends two vectors of lengths `m` and `n` to produce
+one of length `o = m + n`. This is a variant of `Fin.append` with an additional `ho` argument,
+which provides control of definitional equality for the vector length.
+
+This turns out to be helpful when providing simp lemmas to reduce `![a, b, c] n`, and also means
+that `vecAppend ho u v 0` is valid. `Fin.append u v 0` is not valid in this case because there is
+no `Zero (fin (m + n))` instance. -/
+def vecAppend {α : Type _} {o : ℕ} (ho : o = m + n) (u : Fin m → α) (v : Fin n → α) : Fin o → α :=
+  Fin.append u v ∘ Fin.cast ho
+#align matrix.vec_append Matrix.vecAppend
+
+theorem vecAppend_eq_ite {α : Type _} {o : ℕ} (ho : o = m + n) (u : Fin m → α) (v : Fin n → α) :
+    @(vecAppend ho u v) = fun i : Fin o =>
+      if h : (i : ℕ) < m then u ⟨i, h⟩
+      else v ⟨(i : ℕ) - m, (tsub_lt_iff_left (le_of_not_lt h)).2 (ho ▸ i.2)⟩ := by
+  ext i
+  rw [vecAppend, Fin.append, Function.comp_apply, Fin.addCases]
+  congr with hi
+  simp only [eq_rec_constant]
+  rfl
+#align matrix.vec_append_eq_ite Matrix.vecAppend_eq_ite
+
+-- Porting note: proof was `rfl`, so this is no longer a `dsimp`-lemma
```

Comment:
Note that it would be a dsimp lemma if we had

```lean
def Nat.ble : @& Nat → @& Nat → Bool
  | zero,   _      => true
  | succ _, zero   => false
  | succ n, succ m => ble n m
```
instead of
```lean
def Nat.ble : @& Nat → @& Nat → Bool
  | zero,   zero   => true
  | zero,   succ _ => true
  | succ _, zero   => false
  | succ n, succ m => ble n m
```